### DISCUSSION TITLE
when there is a space in the note name, reports an error

### DISCUSSION THREAD

**Discussion by tyf2018** [Taitava/obsidian-shellcommands#155] [Q&A] [answered]
I use the following command to call Eme to open ob's current note, which works fine in most cases, but when there is a space in the note name, the call reports an error. How should I modify the order? Adding double quotes doesn't seem to fix the problem either.

```
C:\totalcmd\Tools\Emeditor\EmEditor.exe {{file_path:absolute}} /l {{caret_position}}
```
Thanks!

**Comment by Taitava** [comment]
Thank you for your question! 🙂 Hmm, as you are using Windows, are you using `CMD` as your shell? [Here are instructions on how to check in the plugin's settings what shell you are using](https://publish.obsidian.md/shellcommands/Operating+systems+and+shells/Shells#How+to+know+which+shell+is+used).

# If you are using `CMD`

If you are using `CMD`, then the issue is that SC does not do any kind of [escaping for special characters](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values) when the shell is CMD. This is because I don't know how to do this kind of escaping in CMD (is it even possible?).

Adding the double quotes around `{{file_path:absolute}}` should help to avoid problems with spaces (although it's still not 100% secure), but as you have already tried them, I wonder why they did not help. Did you use the double quotes like this: `C:\totalcmd\Tools\Emeditor\EmEditor.exe "{{file_path:absolute}}" /l {{caret_position}}` ?

For a more robust solution, I would suggest you to switch `CMD` to `PowerShell`, either `5` or `Core`. I think PowerShell 5 becomes installed on every Windows system. PowerShell Core is newer and needs to be installed manually if you want to use it. For this case, I think whichever of those might do ok. SC escapes special characters in variable values when any of the PowerShells is used.

Switching to PowerShell can cause problems to other shell commands (if you have other commands defined) because it has differently named commands than CMD, but the one you mentioned should work just fine in PowerShell. If you want, instead of changing the shell globally for all the shell commands you have defined in the plugin, you can [change the shell just for this particular shell command](https://publish.obsidian.md/shellcommands/Operating+systems+and+shells/Shells#Choosing+a+shell+on+a+per+shell+command+basis). (Another thing with PowerShell is that it's slower to execute than CMD, it takes a few seconds for shell commands to execute when PowerShell is used.)

If you decide to use PowerShell, then I think (but am not 100 % certain) that you could use the variable without any double quotes.

# If you are already using PowerShell

Then I'd need to think about it more. Can you provide me the error message? You can e.g. select that error output will be inserted into your clipboard, here's a guide (you'll want to change `stderr` to _Clipboard_): https://publish.obsidian.md/shellcommands/Output+handling/Output+channels#How+to+select+output+channels+for+a+shell+command

**Reply by tyf2018** [reply]
Hi， @Taitava 

Thank you very much for such a detailed and patient answer, I learned a lot. Calling Eme has solved the whitespace problem. But now I have another small problem.

I use the following code to call pandoc to convert the notes to docx format. Whether I switch to PowerShell or add double quotes, an error will be reported, so I can only use cmd.

```
"C:/Program Files/Pandoc/pandoc.exe" {{file_path:absolute}} -f markdown -o {{folder_path:relative}}/{{title}}.docx --reference-doc=F:\1Syncthing\1库_剪藏_浏览器\1黑曜石_库\00数据管理_vault\template\1Office_Template\word.docx
```

When switching to PowerShell（For this command, adding double quotes does not work）, the preview screenshot of the command is as follows (excess single quotes will be generated, switch back to cmd, and the single quotes will disappear):
![2022-02-17_10-42-41](https://user-images.githubusercontent.com/56342788/154394529-2807c3fd-9db8-484d-9755-ecd8f468a827.jpg)

The error is as follows:
![2022-02-17_10-58-30](https://user-images.githubusercontent.com/56342788/154396252-354be42a-1a22-4bca-abd2-0ef4d5397537.jpg)

Switch back to cmd, and the command can be used normally (except for spaces in the file name).

Thanks!

**Reply by Taitava** [reply]
Now I'm quite uncertain why this happens, but I think I've seen these � characters appearing when I've used PowerShell 5. I'd really like to find out why it happens some day. Now I'm quite clueless. I guess it's some kind of encoding issue. If you are now using PowerShell 5, you can still try to [install PowerShell Core](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.2) (if you haven't already). And by _Core_ I mean a version of PowerShell that is 6 or above, the current version is 7.2.1.

Please let me know if this didn't help.

**Reply by tyf2018** [reply]
> If you are now using PowerShell 5, you can still try to [install PowerShell Core](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.2) (if you haven't already). And by _Core_ I mean a version of PowerShell that is 6 or above, the current version is 7.2.1.
> 
I installed PowerShell-7.2.1-win-x64.msi, but unfortunately still get the error:

![2022-02-17_16-41-33](https://user-images.githubusercontent.com/56342788/154437869-001edf07-85b0-4908-a595-597439fe2f51.jpg)

**Reply by Taitava** [reply]
Ah, too bad it did not help 🙁 . I'll look into this more and reply when I find something out.

Just to make sure, you probably also did change the shell in the SC settings, too, after installing PS 7? 🙂

> ![2022-02-17_10-42-41](https://user-images.githubusercontent.com/56342788/154394529-2807c3fd-9db8-484d-9755-ecd8f468a827.jpg)

~~Can you copy and paste here the command and also your current file name and folder name as text, please? Maybe I can better test this if I create a folder and a file that uses similar characters as in your language.~~ Edit: nevermind, I just noticed you had the command already in text form, too. 🙂

**Reply by Taitava** [reply]
> When switching to PowerShell（For this command, adding double quotes does not work）, the preview screenshot of the command is as follows (excess single quotes will be generated, switch back to cmd, and the single quotes will disappear):

That's the special characters escaping. `` ` `` gets added in front of every character that is not a letter, a number or underscore `_`. This is to prevent any possible special meanings characters might have and use the character literally. It's a bit hard to explain, but the `` ` `` character appearing in front of special characters is normal. They fix the space issue you had: ` ` becomes `` `  ``. When PowerShell executes your command, it removes the `` ` `` characters. This escaping is only done for values coming from `{{` variables `}}`, not for other things in the command.

**Reply by tyf2018** [reply]
> Just to make sure, you probably also did change the shell in the SC settings, too, after installing PS 7? 🙂

Yes, after installation, I switched to PowerShell Core.
![2022-02-18_09-14-17](https://user-images.githubusercontent.com/56342788/154598704-7de5ee36-fa83-4b39-af80-ab2f6d4c6934.jpg)

> That's the special characters escaping. `` ` `` gets added in front of every character that is not a letter, a number or underscore `_`. This is to prevent……

👍


Finally, I will summarize my usage feedback to avoid misunderstandings.

All my SC commands work fine when the filename does not contain spaces. When spaces are included, the following commands will report an error, and adding double quotes or using PowerShell Core is invalid.

```
"C:/Program Files/Pandoc/pandoc.exe" "{{file_path:absolute}}" -f markdown -o {{folder_path:relative}}/{{title}}.docx --reference-doc=E:\OneDrive\1承接项目\00数据管理\template\1Office_Template\word.docx
```

Best Regards.

**Reply by Taitava** [reply]
Can you try this with PowerShell Core or PowerShell 5:
```PowerShell
Start-Process -FilePath "C:/Program Files/Pandoc/pandoc.exe" -ArgumentList "`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.docx`" --reference-doc=E:\OneDrive\1承接项目\00数据管理\template\1Office_Template\word.docx"
```

If it works, maybe I can then try to describe it's differences. But if not, we can try to think about something else.

**Reply by Taitava** [reply]
Or if that does not work, then maybe with CMD instead of PowerShell:
```CMD
start "" "C:/Program Files/Pandoc/pandoc.exe" "{{file_path:absolute}} -f markdown -o {{folder_path:relative}}/{{title}}.docx --reference-doc=E:\OneDrive\1承接项目\00数据管理\template\1Office_Template\word.docx"
```

But generally, I usually recommend PowerShell over CMD, just for the fact that `{{` variables `}}` are then a bit more safer to use. But if PowerShell does not work at all, then of course switch back to CMD.

**Reply by tyf2018** [reply]
> Can you try this with PowerShell Core or PowerShell 5:
> 
> ```powershell
> Start-Process -FilePath "C:/Program Files/Pandoc/pandoc.exe" -ArgumentList "`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.docx`" --reference-doc=E:\OneDrive\1承接项目\00数据管理\template\1Office_Template\word.docx"
> ```

Thank you so much, the problem is solved!

I have to say, your patient and detailed answer even moved me.

Best Regards.

**Reply by Taitava** [reply]
Thank you! ❤️ 

These issues are really good for me, as I learn new things about PowerShell and shells in general when I try out things myself and research different problems.

> If it works, maybe I can then try to describe it's differences.

Just for the record, this description I promised can be found [below in Solution 1](https://github.com/Taitava/obsidian-shellcommands/discussions/155#discussioncomment-2213135), although it's for a different command.

----

One thing that I learned, was about the `�` characters that appeared in some of your error messages:

> ![2022-02-17_16-41-33](https://user-images.githubusercontent.com/56342788/154437869-001edf07-85b0-4908-a595-597439fe2f51.jpg)

At first I thought that there would be an issue with an _input_ encoding, which would then lead to PowerShell not understanding the command. But now I guess the input encoding might actually be just ok, but a problem with _output_ encoding is preventing the error message from displaying the "corrupted" `�` symbols correctly. I hope that this is just a cosmetic issue for your use case. I guess and hope it does not affect the output of Pandoc. Anyway, I will try to find a solution to this `�` output issue in discussion #157.

**Comment by tyf2018** [comment, marked as answer]
But I am a little confused, the same format (exported to html), the following command will prompt the command to fail:

```
Start-Process -FilePath "C:/Program Files/Pandoc/pandoc.exe" -ArgumentList "`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.html`"
```

**Reply by Taitava** [reply]
# Solution 1

It's a really tiny error that is hard to spot: There's one double quote `"` missing from the end. This should work:
```PowerShell
Start-Process -FilePath "C:/Program Files/Pandoc/pandoc.exe" -ArgumentList "`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.html`""
```

(I'm learning constantly new things about PowerShell when inspecting these issues 🙂. While the above solution does work, a cleaner solution will be at the bottom of this message, Solution 2. As I just learned the cleanest solution, I still like to include a middle step that I went through before finding the best solution.)

I'll try to go through how the double quotes work in the above shell command:
1. `Start-Process` is given just two arguments: `-FilePath` and `-ArgumentList`.
2. `-FilePath` is quite simple as it's just `"C:/Program Files/Pandoc/pandoc.exe"`.
3. `-ArgumentList` is more complex, because at the beginning it's a single argument, but which will become a set of three arguments: `"{{file_path:absolute}}"`, `-f markdown` and `-o "{{folder_path:relative}}/{{title}}.html"`. Two of these arguments are wrapped in double quotes due to spaces (which I later noticed is not needed, but let's keep the double quotes for now, just for the sake of learning new things 🙂 ). When these three arguments need to be concatenated into a single argument for `-ArgumentList`, we need to wrap all the three arguments under the same pair of double quotes: `-ArgumentList ""{{file_path:absolute}}" -f markdown -o "{{folder_path:relative}}/{{title}}.html""`, but this becomes invalid, because like mentioned, two arguments uses double quotes already. For this to work, we need to _escape_ all the _inner_ double quotes that are in between the _outer_ double quotes. The inner double quotes will have a `` ` `` preceding them, while the outer double quotes will not have this kind of prefixes: ``-ArgumentList "`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.html`""``.
4. When `Start-Process` command runs, it will execute `"C:/Program Files/Pandoc/pandoc.exe"` and give it the arguments it received, but the form of the arguments will change from ``"`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.html`""`` to a more normal `"{{file_path:absolute}}" -f markdown -o "{{folder_path:relative}}/{{title}}.html"`.

P.S. When we think about this shell command ending with two double quotes `` `"" ``, we might ask why there were no two double quotes needed in the previous shell command:
>```PowerShell
>Start-Process -FilePath "C:/Program Files/Pandoc/pandoc.exe" -ArgumentList "`"{{file_path:absolute}}`" -f markdown -o `"{{folder_path:relative}}/{{title}}.docx`" --reference-doc=E:\OneDrive\1承接项目\00数据管理\template\1Office_Template\word.docx"
>```

That's because this previous shell command ends with an argument that does not use double quotes at all: `--reference-doc=E:\OneDrive\1承接项目\00数据管理\template\1Office_Template\word.docx`.

I hope I was able to explain this in a way that it makes some sense.

----

# Solution 2

We don't need to use `Start-Process` at all. I used it because I didn't know any better way how to make calling `"C:/Program Files/Pandoc/pandoc.exe"` to work.

I just learned that this one does not work...
```PowerShell
"C:\Program Files\Pandoc\pandoc.exe" "{{file_path:absolute}}" -f markdown -o "{{folder_path:relative}}\{{title}}.html"
```

...because PowerShell does not understand that the first item is a callable program. PS does not understand it because it's wrapped in quotes (due to the space, of course). But we can fix it with a tiny trick:

```PowerShell
C:\Program` Files\Pandoc\pandoc.exe "{{file_path:absolute}}" -f markdown -o "{{folder_path:relative}}\{{title}}.html"
```

So the space ` `  is escaped by making it to be `` `  ``.

We can even remove all double quotes:

```PowerShell
C:\Program` Files\Pandoc\pandoc.exe {{file_path:absolute}} -f markdown -o {{folder_path:relative}}\{{title}}.html
```

This still works even for files that have spaces in their names, because now that PowerShell is used, SC escapes the spaces with `` `  ``.

P.S. I changed all forward slashes `/` to backward slashes `\` as that's how Windows usually presents path delimiters. But I did not notice any issues when using `/` instead of `\` so maybe `/` can be used as path delimiters, too, but I just wanted to change them just in case.

**Comment by Taitava** [comment]
I changed the category from _Debugging and testing_ to _Q & A_.
